src/wallet/wallet.cpp

```diff
@@ -4339,6 +4340,73 @@ bool CWallet::BackupWallet(const std::string& strDest) const
     return GetDatabase().Backup(strDest);
 }
 
+bool CWallet::QueuedTransactionExists(const uint256& txid) const
+{
+    AssertLockHeld(cs_wallet);
+    return queuedTransactionMap.count(txid) > 0;
+}
+
+bool CWallet::WriteQueuedTransaction(const uint256 &txid, const CMutableTransaction &tx)
+{
+    AssertLockHeld(cs_wallet);
+
+    const bool success = WalletBatch(GetDatabase()).WriteQueuedTransaction(txid, tx);
+    if(success)
+        queuedTransactionMap[txid] = tx;
+
+    return success;
+}
+
+bool CWallet::EraseQueuedTransaction(const uint256& txid)
+{
+    AssertLockHeld(cs_wallet);
+    const bool success = WalletBatch(GetDatabase()).EraseQueuedTransaction(txid);
+    if(success)
+        queuedTransactionMap.erase(txid);
+    return success;
+}
+
+bool CWallet::GetQueuedTransaction(const uint256 &txid, CMutableTransaction *data) const
+{
+    AssertLockHeld(cs_wallet);
+    auto it = queuedTransactionMap.find(txid);
+    if (it == queuedTransactionMap.end())
+        return false;
+    if (data != nullptr)
+        (*data) = it->second;
+    return true;
+}
+
+void CWallet::EffectTransactionQueue()
+{
+    // This is called when we get a new block to deal with queued transactions.
+    if (chain().isInitialBlockDownload())
+        return;
+    // We only do this for fresh blocks. Otherwise, we might trigger way too early.
+    std::set<uint256> to_dequeue;
+    // Kludge; we can't remove items in a map while iterating over it.
+    AssertLockHeld(cs_wallet);
+    for (const auto& i : queuedTransactionMap)
+    // For each transaction in the queue...
+    {
+        const uint256& txid = i.first;
+        const CMutableTransaction& tx = i.second;
+
+        std::string unused_err_string;
+        if (chain().broadcastTransaction(MakeTransactionRef(tx), m_default_max_tx_fee, /* relay */ true, unused_err_string))
```

Comment:
What happens if a transaction on the queue is directly included in a block later on?  From reading the code, it seems that `broadcastTransaction` would fail in that case, so the transaction would stay on the queue forever?  Even if that's not the case, I think there should be a test for this situation as well (it can probably be done relatively easily with `invalidateblock` and `reconsiderblock` to get a temporary fork between two nodes and then let the other node (which has the tx queued) sync up to the full chain).

Maybe it works already "accidentally" if `EffectTransactionQueue` is called block-after-block, which means that the node will try to broadcast right when the tx becomes valid, before knowing it was already confirmed in a later block.  But that is certainly not guaranteed, as e.g. the node could consider itself in IBD still at that point.

Reply:
It would stay there forever. This is the intended behavior.

The alternative would be to try and clean it out in `EffectTransactionQueue`, but I'm not sure how to do this (we only have a limited API that returns true/false, not full validation). Also, I'm not sure if it's desired. We may wish to queue up a transaction whose inputs are already spent, in the hopes that a reorg will take them out and make way for our transaction.

Do you see a simple way to add all these checks to `EffectTransactionQueue`? The validity check was just a small convenience thing, I don't think the API here is supposed to be *that* user-friendly (i.e. there is no max fee limit as there is in sendrawtransaction). In my opinion, it doesn't really hurt to have transactions floating around like that.

Reply:
Ok I see, so you think this is actually working-as-intended (because of reorgs) / we are fine with the tx staying on the queue and perhaps being manually cleaned out from it later.  That's fine with me then.